### PR TITLE
Add LazyInsertions

### LINKED ISSUE #213: Lazy Insertion
Constructing an entity while the world is running isn't very ergonomic right now and the naive way of doing it can lead to almost no parallel systems (Accessing each storage required for construction mutably, so it blocks all other systems using that storage). A better approach is something like an "initialization" system where it is possible to insert a flag component that signifies that a certain component should be inserted. For example:
```rust
struct Position { x: f32, y: f32, z: f32 };
struct InitPosition;
struct InitPositionSystem;
impl<'a> System<'a> for InitPositionSystem {
    type SystemData = (
        Entities<'a>, 
        WriteStorage<'a, Position>, 
        ReadStorage<'a, InitPosition>
    );
    fn run(&mut self, (entities, positions, inits): Self::SystemData) {
        for (entity, _) in (&*entities, &inits).join() {
            positions.insert(entity, Position { x: 0.0, y: 0.0, z: 0.0 });
        }
    }
}
```

I feel like this could be abstracted over a bit by allowing some sort of lazy insertion method for the world. So instead of having to set up all this for an initialization system, you could just do something like:
```rust
impl<'a> System<'a> for SomeSys {
    type SystemData = (
        LazyInsert<'a>,
        Entities<'a>, 
    );
    fn run(&mut self, (lazy, entities): Self::SystemData) {
        for entity in (&*entities).join() {
            if ... { // some condition to insertion or something
                lazy.insert::<Position>(entity, Position { x: 0.0, y: 0.0, z: 0.0 });
            }
        }
    }
}
```
This would put the component into some queue in the world, so then once the world is maintained it inserts the components.

### PR DESCRIPTION
**Pull request by torkleyy**
Fixes #213

### CODE REVIEW

File: src/world.rs [hunk 1]
```diff
@@ -366,6 +366,60 @@ impl Generation {
...
+}
+
+/// Lazy insertions can be used after creating a new
+/// entity in a system. This way, none of the actual
+/// component storages have to be borrowed mutably.
+///
+/// This resource is added to the world by default.
+#[derive(Default)]
+pub struct LazyInsertions {
+    vec: Vec<Box<LazyInsertInternal>>
```
Comment (torkleyy) [inline src/world.rs]: Should use an atomic linked list.

File: src/world.rs [hunk 2]
```diff
@@ -366,6 +366,60 @@ impl Generation {
...
+
+impl<L> LazyInsertInternal for Option<L>
+    where L: LazyInsert
+{
+    fn insert(&mut self, world: &World) {
+        self.take().unwrap().insert(world);
+    }
+}
+
+/// Lazy insertions can be used after creating a new
```
Comment (torkleyy) [inline src/world.rs]: Needs docs somewhere that the insertions are in reverse order.

File: src/world.rs [hunk 3]
```diff
@@ -366,6 +366,60 @@ impl Generation {
...
+///
+/// This resource is added to the world by default.
+#[derive(Default)]
+pub struct LazyInsertions {
+    vec: Vec<Box<LazyInsertInternal>>
+}
+
+impl LazyInsertions {
+    /// Adds an insertion.
+    pub fn add<L>(&mut self, l: L)
```
Comment (torkleyy) [inline src/world.rs]: Missing usage example

**Comment by Aceeri** [comment]
https://github.com/slide-rs/specs/issues/213

File: src/world.rs [hunk 4]
```diff
@@ -366,6 +366,60 @@ impl Generation {
...
+
+trait LazyInsertInternal: Send + Sync {
+    fn insert(&mut self, world: &World);
+}
+
+impl<L> LazyInsertInternal for Option<L>
+    where L: LazyInsert
+{
+    fn insert(&mut self, world: &World) {
+        self.take().unwrap().insert(world);
```
Comment (kvark) [inline src/world.rs]: should we check for `None` here?

Comment (torkleyy) [inline src/world.rs]: It's local to the module, so cannot be triggered by the user. I doubt that `expect` or some special casing would be of much use here.

File: src/world.rs [hunk 5]
```diff
@@ -366,6 +367,66 @@ impl Generation {
...
+{
+    fn insert(self, world: &World) {
+        for item in self {
+            item.insert(world);
+        }
+    }
+}
+
+trait LazyInsertInternal: Send + Sync {
+    fn insert(&mut self, world: &World);
```
Comment (WaDelma) [inline src/world.rs]: I am slightly confused why doesn't this method take ownership of `self` and instead uses `Option::take`.

Comment (torkleyy) [inline src/world.rs]: That's because `Box<Foo>` cannot be the receiver of a self-moving method since `self` is `!Sized`.

**Comment by kvark** [comment]
@torkleyy r=me when ready

File: src/world.rs (hunk 5 above)
Comment (WaDelma) [inline src/world.rs]: Hmmm... Can't you take it as `Box<self>` then?

Comment (WaDelma) [inline src/world.rs]: Hmm... Can't you take it as `self: Box<Self>` then?

Comment (torkleyy) [inline src/world.rs]: I tried that first but it didn't work out at some point. Well, I found another solution using `Box<Self>`.

**Comment by kvark** [comment]
Might be a good idea to squash the commits a bit, otherwise all looking good.

**Pull request merged**